File: CPP/Algorithms/SortSearch/DFSadjList.cpp
// C++ implementation of the approach
#include <bits/stdc++.h>
using namespace std;


void addEdge(vector<int> adj[], int s, int d) {
  adj[s].push_back(d);
  adj[d].push_back(s);
}

void dfs(vector<int> adj[], int start, vector<bool>& visited)
{
	cout << start << " ";
	visited[start] = true;

	for (int i = 0; i < adj[start].size(); i++) {

    int node = adj[start][i]; // all adjacency node of start
		if (!visited[node]) {
			dfs(adj, node, visited);
		}
	}
}


void printGraph(vector<int> adj[], int V) {
  for (int d = 0; d < V; ++d) {
    cout << "\n Vertex "
       << d << ":";
    for (auto x : adj[d])
      cout << "-> " << x;
    printf("\n");
  }
}

int main()
{

	int V = 5;
  vector<int> adj[V];

	addEdge(adj, 0, 1);
	addEdge(adj, 0, 2);
	addEdge(adj, 0, 3);
	addEdge(adj, 0, 4);

	vector<bool> visited(V, false);

  printGraph(adj, V);
  cout<<endl;
  
	dfs(adj, 0, visited);
}

File: JavaScript/Algorithms/SortSearch/DFSadjList.js
class Graph
{
	constructor(v)
	{
		this.V = v;
		this.adj = new Array(v);
		for(let i = 0; i < v; i++)
			this.adj[i] = [];
	}
	

	addEdge(v, w)
	{
		this.adj[v].push(w); 
	}
	
	// A function used by DFS
	DFSUtil(v, visited)
	{
		
		// Mark the current node as visited and print it
		visited[v] = true;
		console.log(v + " ");

		// Recur for all the vertices adjacent to this
		// vertex
		for(let i of this.adj[v].values())
		{
			let n = i
			if (!visited[n])
				this.DFSUtil(n, visited);
		}
	}

	DFS(v)
	{

		let visited = new Array(this.V);
		for(let i = 0; i < this.V; i++)
			visited[i] = false;

		this.DFSUtil(v, visited);
	}
}


g = new Graph(4);

g.addEdge(0, 1);
g.addEdge(0, 2);
g.addEdge(1, 2);
g.addEdge(2, 0);
g.addEdge(2, 3);
g.addEdge(3, 3);

console.log("Following is Depth First Traversal " + 
			"(starting from vertex 2)");

g.DFS(2);

